File: CodeForces/scanners/scanner.swift
import Foundation

class Scanner {
    private var buffer: [String] = []

    private func refill() {
        while buffer.isEmpty {
            guard let line = readLine() else {
                fatalError("EOF reached")
            }
            buffer = line.split(separator: " ").map { String($0) }
        }
    }

    func nextInt() -> Int {
        refill()
        return Int(buffer.removeFirst())!
    }

    func nextDouble() -> Double {
        refill()
        return Double(buffer.removeFirst())!
    }

    func nextLong() -> Int64 {
        refill()
        return Int64(buffer.removeFirst())!
    }

    func nextWord() -> String {
        refill()
        return buffer.removeFirst()
    }

    func nextString() -> String {
        refill()
        let line = buffer.joined(separator: " ")
        buffer.removeAll()
        return line
    }

    func nextIntArray(sorted: Bool = false) -> [Int] {
        refill()
        let arr = buffer.map { Int($0)! }
        buffer.removeAll()
        return sorted ? arr.sorted() : arr
    }

    func nextLongArray(sorted: Bool = false) -> [Int64] {
        refill()
        let arr = buffer.map { Int64($0)! }
        buffer.removeAll()
        return sorted ? arr.sorted() : arr
    }
}

File: CodeForces/scanners/scanner.rs
use std::io::{self, BufRead};
use std::str::FromStr;

struct Scanner<R> {
    reader: R,
    buffer: Vec<String>,
}

impl<R: BufRead> Scanner<R> {
    fn new(reader: R) -> Self {
        Scanner {
            reader,
            buffer: Vec::new(),
        }
    }

    fn refill(&mut self) {
        while self.buffer.is_empty() {
            let mut line = String::new();
            let n = self.reader.read_line(&mut line).expect("Failed to read line");
            if n == 0 {
                panic!("EOF reached");
            }
            self.buffer = line
                .split_whitespace()
                .map(|s| s.to_string())
                .collect();
        }
    }

    fn next<T: FromStr>(&mut self) -> T
    where
        T::Err: std::fmt::Debug,
    {
        self.refill();
        self.buffer
            .remove(0)
            .parse::<T>()
            .expect("Failed to parse token")
    }

    fn nextInt(&mut self) -> i32 {
        self.next()
    }

    fn nextDouble(&mut self) -> f64 {
        self.next()
    }

    fn nextLong(&mut self) -> i64 {
        self.next()
    }

    fn nextWord(&mut self) -> String {
        self.next()
    }

    fn nextString(&mut self) -> String {
        self.refill();
        let line = self.buffer.join(" ");
        self.buffer.clear();
        return line;
    }

    fn nextIntArray(&mut self, sorted: bool) -> Vec<i32> {
        self.refill();
        let mut arr: Vec<i32> = self.buffer.drain(..).map(|s| s.parse().unwrap()).collect();
        if sorted {
            arr.sort();
        }
        arr
    }

    fn nextLongArray(&mut self, sorted: bool) -> Vec<i64> {
        self.refill();
        let mut arr: Vec<i64> = self.buffer.drain(..).map(|s| s.parse().unwrap()).collect();
        if sorted {
            arr.sort();
        }
        arr
    }
}
